[v3] Refactor: Cut Redundant Dependency in order to use as Standalone

sasezaki opened the pull request:
## Refactor: Cut Redundant Dependency in order to use as Standalone.

When using Paginator class as Dependency, there is problem.
- A lot of method is defined, regardless of whether or not to use.
  - json, filter, render, cache 
- ScrollingStyle is static dependecy.
  - it's not need when ScrollingStyle object is passed.
    - eg. `$paginator->getPages(new Zend\Paginator\ScrollingStyle\All()))`

Above thing is a headache when to use standalone.

So I request refactoring this component.

Introduced below interface, trait, class.
- PaginatorInterface
- JsonSerializeTrait, RenderTrait, FilterTrait, CachedTrait
- SimplePaginator
- GlobalPaginator

.. And Paginator class can implements like below, without BC break.

``` php
class Paginator extends GlobalPaginator
{
    use JsonSerializeTrait,
        RenderTrait,
        FilterTrait,
        CachedTrait;
```

Only ScrollingStyleInterface implentation must be changed. 
BEFORE  `public function getPages(Paginator $paginator, $pageRange = null);`
AFTER  `public function getPages(PaginatorInterface $paginator, $pageRange = null);`

sasezaki committed:
Move toJson() method To JsonSerializeTrait

sasezaki committed:
Move render()  & __toString() requires To RenderTrait

sasezaki committed:
Move caching To CachedTrait

todo getItemsByPage() must override by caching

sasezaki committed:
Introduce PaginatorInterface

Introduced SimplePaginator, too.
Cache & Filter is required getItemsByPage

sasezaki committed:
oops, missing CachedTrait

sasezaki committed:
Move Global (static access) to GlobalPaginator

sasezaki committed:
fix namespace requirement

sasezaki committed:
remove duplicate method

sasezaki committed:
fix namespace import

sasezaki committed:
default itemCountPerPage must be null, when global count exists

getItemCountPerPage() works with static::getDefaultItemCountPerPage(),
lazy count-per-page check

sasezaki committed:
Add doc-comment in JsonSerializeTrait::jsonSerialize()

sasezaki committed:
move stdlib dependency as dev

Stdlib only require ArrayUtils::iteratorToArray() method.
It used in GlobalPaginator::setGlobalConfig.

sasezaki committed:
TypeHint must be PaginatorInterface

sasezaki committed:
Apply php-cs-fixer

fix invalid lines, remove useless `use`

marc-mabe commented:
@sasezaki the caching should also be moved into the adapters because the adapter only know how to cache paginater results. I think the issues for that are in the zf2 repo.

marc-mabe commented:
@sasezaki the caching should also be moved into the adapters because the adapter only know how to cache paginater results. I think the issues for that are in the zf2 repo.

Ocramius commented:
Caching could still exist, but on a separate instance. Agreed tho: it needs to die for now (the current approach is using static caching, that can lead to horrible SNAFU moments)

sasezaki commented:
To be clear, talking cache problem is current v2 Paginator implemntaions ?
I agree caching problem.

Firstly, this PR only tried cutting Redundant Dependency for current implementaion.
Should I continue update on this PR as complete update for v3 ?

marc-mabe commented:
@sasezaki @Ocramius I agree that the caching issue is not part of this PR but it should be addressed in another PR before v3.

sasezaki committed:
remove jsonSerialize()

Don't add new method when Refactoring.
(oops, jsonSerialize() introduced with my misunderstood usage.)

weierophinney commented:
@sasezaki Could you provide some usage examples? These will be important for understanding the impact to end users, and will also shed more light on the design changes.

Thanks!

samsonasik commented on src/CachedTrait.php:
```diff
@@ -0,0 +1,147 @@
+    protected function _getCacheInternalId()
+    {
+        return md5(serialize([
+            spl_object_hash($this->getAdapter()),
```
this cache internal id fixed by https://github.com/zendframework/zend-paginator/pull/33 , this may need copy function from current master branch.

weierophinney commented:
This repository has been closed and moved to laminas/laminas-paginator; a new issue has been opened at https://github.com/laminas/laminas-paginator/issues/5.

weierophinney commented:
This repository has been moved to laminas/laminas-paginator. If you feel that this patch is still relevant, please re-open against that repository, and reference this issue. To re-open, we suggest the following workflow:
 - Squash all commits in your branch (`git rebase -i origin/{branch}`)
 - Make a note of all changed files (`git diff --name-only origin/{branch}...HEAD
 - Run the laminas/laminas-migration tool on the code.
 - Clone laminas/laminas-paginator to another directory.
 - Copy the files from the second bullet point to the clone of laminas/laminas-paginator.
 - In your clone of laminas/laminas-paginator, commit the files, push to your fork, and open the new PR.
 We will be providing tooling via laminas/laminas-migration soon to help  automate the process.
